core/src/server/rpc_module.rs

```diff
@@ -762,34 +834,42 @@ impl<Context: Send + Sync + 'static> RpcModule<Context> {
 		let callback = {
 			self.methods.verify_and_insert(
 				subscribe_method_name,
-				MethodCallback::new_subscription(Arc::new(move |id, params, method_sink, conn, claimed| {
+				MethodCallback::new_subscription(Arc::new(move |id, params, method_sink, conn| {
 					let uniq_sub = SubscriptionKey { conn_id: conn.conn_id, sub_id: conn.id_provider.next_id() };
 
 					// response to the subscription call.
 					let (tx, rx) = oneshot::channel();
 
-					let sink = SubscriptionSink {
+					let sink = PendingSubscriptionSink {
 						inner: method_sink,
-						close_notify: Some(conn.close_notify),
 						method: notif_method_name,
 						subscribers: subscribers.clone(),
 						uniq_sub,
-						id: Some((id.clone().into_owned(), tx)),
-						unsubscribe: None,
-						_claimed: claimed,
+						id: id.clone().into_owned(),
+						subscribe: tx,
 					};
 
-					// The callback returns a `SubscriptionResult` for better ergonomics and is not propagated further.
-					if callback(params, sink, ctx.clone()).is_err() {
-						tracing::warn!("Subscribe call `{}` failed", subscribe_method_name);
-					}
+					// The subscription callback is a future from the subscription
+					// definition and not the as same when the subscription call has been completed.
+					//
+					// This runs until the subscription callback has completed.
+					let sub_fut = callback(params.into_owned(), sink, ctx.clone());
+
+					tokio::spawn(async move {
+						if let Err(SubscriptionCallbackError::Some(msg)) = sub_fut.await {
+							tracing::warn!("Subscribe call `{subscribe_method_name}` failed: {msg}");
+						}
```

Comment:
I guess this makes sense to me; I kept mulling over what should happen if a user returns `Err(e)` in a subscription thing now, and the options are to either return it to the user as a subscription error, or not.

The nice thing about not returning the error to the user is that if you want to return some error to the user you have to do it super explicitly, otherwise no error is returned. (The downside is that if they return an error from the callback, they might expect it to be sent to the user but it won't be?).

On the other hand, if you _did_ send any error to the user here, you could remove APIs like `PendingSubscriptionSink::reject()` and `SubscriptionSink::close(error)`, and instead, returning from the subscription async callback with `Ok` would mean it closed successfully and no need to send an error, and returning with `Err(e)` means it closes and to send an error back in whatever form makes sense.

Just thinking out loud though; I guess I don't like that the error is silently disappeared from the client perspective at present, but I do like that the current approach is very explicit about things, so I'm happy either way and this doesn't block merging for me!

Reply:
we do send an error to the "client" if _subscription call_ fails i.e, the user wouldn't call `accept/reject`.

when we get back an error here it's just some special logic from the user code that failed such as `SubscriptionMessage` or `Sink::send_timeout` or whatever.

what I am trying to say if a subscription just executes "correctly" or is "dropped" because of some error in the async closure then the client wouldn't be notified unless one calls `SubscriptionSink::close.